### ISSUE TITLE
Atomics used in coroutines are ~2x slower than AtomicReference on Android on write

### ISSUE THREAD

**Issue by ShikaSD** [Kotlin/kotlinx.coroutines#3950]
The coroutines are using atomic reference implementation backed by `Atomic*FieldUpdater` which is 2x slower for `compareAndSet` and `set` when compared to `AtomicReference` on Android devices.

Running the [benchmark](https://github.com/ShikaSD/benchmark-playground/blob/main/benchmark/src/androidTest/java/com/example/compose/benchmark/AtomicReferenceBenchmark.kt) on Pixel 4a I see the following results:

```
 29.0 ns  atomicReference_getAndSwap
 71.3 ns  atomicRef_getAndSwap
 50.7 ns  atomicReference_compareAndSet
135   ns  atomicRef_compareAndSet
  4.3 ns  atomicReference_get
  4.2 ns  atomicRef_get
 18.0 ns  atomicReference_lazySet
 79.4 ns  atomicRef_lazySet
```

In the benchmark above, `AtomicReference` is 2x to 4x faster on write than `atomicfu`. Looking at method traces from Compose benchmarks, I see that certain `Job` and `CoroutineContext` operations run multiple atomic operations depending on `Job` graph complexity.

You can see the trace from one of the abovementioned benchmarks [here](https://ui.perfetto.dev/#!/?s=9114767b39f92fb3dac2bd4162ada6c237157f3128d419c590b2a577b101c5c2) (it should be focused on coroutine work). It does not represent exact timing due to overhead of tracing every method, but does represent the amount of work being done. Most of it is dominated by `Atomic*FieldUpdater` operations and class instance checks that `Atomic*FieldUpdater` does.

As an experiment, I forked `atomicfu` to be backed by `AtomicReference` and got ~5% improvement in the same Compose benchmark, which is pretty significant given amount work Compose executes outside of coroutine context.

I filed a separate issue for Android internally to potentially fix this on the runtime side, but any fix will affect only relatively small subset of modern devices that support runtime updates.

**Comment by ShikaSD** [comment]
It looks like atomicfu supports [generating `VarHandle` based](https://github.com/Kotlin/kotlinx-atomicfu/blob/master/atomicfu-gradle-plugin/src/main/kotlin/kotlinx/atomicfu/plugin/gradle/AtomicFUGradlePlugin.kt#L475) references (which are used for `AtomicReference` implementation in Android). This implementation should be significantly faster than current `Atomic*FieldUpdater` one based on the benchmarks above.

Would it be possible to enable it by default for the jvm Android target (either by configuring kotlinx.coroutines build or by updating the default in atomicfu)?

**Comment by dkhalanskyjb** [comment]
`VarHandle` was introduced in Java 9, but this library supports Java 8. Also, there's no "JVM Android target", just the JVM target, which Android also uses.

**Comment by dkhalanskyjb** [comment]
Now that I think about it, maybe we could make a multi-release jar file that contained both the Java 8 implementation and the Java 9 one.

**Comment by ShikaSD** [comment]
My suggestion was to maybe introduce a separate Android target, but I guess it is not going to work for consumers that ship JVM-only as well. I wonder how well Android supports VarHandle going back a few versions now though, assuming limited support even for Java 8 at ~API 21, so maybe my assumption was incorrect.

**Comment by JakeWharton** [comment]
`VarHandle` is only supported on [API 33](https://developer.android.com/reference/java/lang/invoke/VarHandle) which is basically 0% of devices. Beyond that, it actually requires a new bytecode which will be rejected by older runtimes. This means you cannot even conditionally reference them as you would a regular new API (such as java.time, for example). So it will actually require a minimum SDK of 33 (or maybe 29 actually at which point you then could conditionally reference it). Android lacks a version of MR jars which would allow selecting an entirely new dex file based on runtime support, sadly.

**Comment by ShikaSD** [comment]
Right, disregard my last suggestion then. I am still concerned that AtomicReference performance is way better though, so probably worth investigating how it works on previous versions of Android.

**Comment by ShikaSD** [comment]
As I suspected, older versions of Android [use `Unsafe`](https://cs.android.com/android/_/android/platform/libcore/+/dedcf311920a432c4a5fbc32dab60bc749c461ff:ojluni/src/main/java/java/util/concurrent/atomic/AtomicReference.java;bpv=1;bpt=0;drc=cc42201b8d47d4c03a69dcf6df4e58c1a11d32e5;drf=luni%2Fsrc%2Fmain%2Fjava%2Fjava%2Futil%2Fconcurrent%2Fatomic%2FAtomicReference.java) to perform these operations.

**Comment by qwwdfsad** [comment]
What is possible to do is to release a separate variant of the `kotlinx-coroutines-core` where transformations are not applied (it's a bit trickier probably because non-transformed code depends on `atomicfu` library), but we definitely do not want to make it the default version -- as it will be slower than the baseline on the rest of the targets

**Comment by ShikaSD** [comment]
I don't think the version without transformation is going to help here, as `atomicfu` implements references the same way by default.

**Comment by ShikaSD** [comment]
This is now being addressed by R8 and ART, so this issue is no longer relevant.

**Comment by ApoloApps** [comment]
this should be reconsidered since it shouldn't be R8 the only way performance optimizations can be made in kotlin libraries
